src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -164,9 +170,127 @@ class CryptoKeyService(Service):
+                CryptoKeyService.EXTENSIONS[attr] = inspect.getfullargspec(attr_obj.__init__).args[1:]
+
+        return CryptoKeyService.EXTENSIONS
+
+    def add_extensions(self, cert, extensions_data, key, issuer=None):
+        # issuer must be a certificate object
+        # By default we add the following
+        cert = cert.public_key(
+            key.public_key()
+        ).add_extension(
+            x509.SubjectKeyIdentifier.from_public_key(key.public_key()), False
+        )
+
+        for extension in filter(lambda v: v[1]['enabled'], extensions_data.items()):
+            klass = getattr(x509.extensions, extension[0])
+            cert = cert.add_extension(
+                klass(*self.get_extension_params(extension, cert, issuer)),
+                extension[1].get('extension_critical') or False
+            )
+
+        return cert
+
+    def get_extension_params(self, extension, cert=None, issuer=None):
+        params = []
+
+        if extension[0] == 'BasicConstraints':
+            params = [extension[1].get('ca'), extension[1].get('path_length')]
+        elif extension[0] == 'ExtendedKeyUsage':
+            usages = []
+            for ext_usage in extension[1].get('usages', []):
+                usages.append(getattr(x509.oid.ExtendedKeyUsageOID, ext_usage))
+            params = [usages]
+        elif extension[0] == 'KeyUsage':
+            params = [extension[1].get(k, False) for k in self.extensions()['KeyUsage']]
+        elif extension[0] == 'AuthorityKeyIdentifier':
+            params = [
+                x509.SubjectKeyIdentifier.from_public_key(
+                    issuer.public_key() if issuer else cert._public_key
+                ).digest if cert or issuer else None,
+                None, None
+            ]
+
+            if extension[1]['authority_cert_issuer'] and cert:
+                params[1:] = [
+                    [x509.DirectoryName(cert._issuer_name)],
+                    issuer.serial_number if issuer else cert._serial_number
+                ]
+
+        return params
+
+    @accepts(
+        Ref('cert_extensions'),
+        Str('schema')
+    )
+    def validate_extensions(self, extensions_data, schema):
+        # We do not need to validate some extensions like `AuthorityKeyIdentifier`.
+        # They are generated from the cert/ca's public key contents. So we skip these.
+
+        skip_extension = ['AuthorityKeyIdentifier']
+        verrors = ValidationErrors()
+
+        for extension in filter(
+            lambda v: v[1]['enabled'] and v[0] not in skip_extension,
+            extensions_data.items()
+        ):
+            klass = getattr(x509.extensions, extension[0])
+            try:
+                klass(*self.get_extension_params(extension))
+            except Exception as e:
+                verrors.add(
+                    f'{schema}.{extension[0]}',
+                    f'Please provide valid values for {extension[0]}: {e}'
```

Comment:
```suggestion
                    f'Please provide valid values for {extension[0]}: {e}'
```
This could use an example so the user will have a reference to what values are acceptable.

Reply:
@william-gr thoughts on this one ? Reason this has been done this way @Mrt134 is that we can in future support more certificate extensions and it doesn't seem right to validate all of them on our own and especially when it is dynamic for some of them ( values differing based on other values provided to the same extension ), so if we handle each extension validation ourselves apart from schema, that would be a lot of unnecessary code to maintain in my opinion. That however would be the only way to make the error message here more elaborate wrt valid values for each extension.

Reply:
If I understand correctly this will hardly be ever seen by the UI users, this is advanced usage of the API, people using these manually should now before hand what they are supposed to be.
IMO its fine to leave it like that.